### PR TITLE
C++ gradient for Select

### LINKED ISSUE #14845: No gradient defined for op: Select
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 x64
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4
- **Python version**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: -

### Describe the problem
It seems there is no gradient defined for the Select operation in the C++ API.

I am actually getting this issue while using the C++ API through the C# bindings provided by the TensorFlowSharp project, and for this reason I didn't fill the "exact command to reproduce" field above. However, seeing that [```@ops.RegisterGradient("Select")``` is placed in math.grad.py](https://github.com/tensorflow/tensorflow/blob/27767d8e9c1325979cf32ff5b81c10df9006fd57/tensorflow/python/ops/math_grad.py#L919), and given that there is no analogous ```REGISTER_GRADIENT_OP("Select", SelectGrad)``` instruction in [math_grad.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/gradients/math_grad.cc), tells me that the gradient for the ```Select``` op is indeed still missing from the C++ API.

Hope its not a false alarm given that I didn't test TensorFlow's .dll directly.

### PR DESCRIPTION
**Pull request by facaiy**
Fix #14845 

migrate python implementation to c++ side, source: https://github.com/tensorflow/tensorflow/blob/27767d8e9c1325979cf32ff5b81c10df9006fd57/tensorflow/python/ops/math_grad.py#L919

### How to test

+ [x] add test case
+ [ ] pass all tests

### CODE REVIEW

**Comment by facaiy** [comment]
@suharshs Hi, since `Select` seems renamed to `Where3`, I'm a little confused about which op the gradient should be registered for, `Select` or `Where3`? Could you give a advice? Thank you.

**Comment by suharshs** [comment]
Sorry for the delay, was on vacation :) 

I believe at the kernel level, the op is still "Select" so you should register it for that.

File: tensorflow/cc/gradients/math_grad.cc [hunk 1]
```diff
@@ -728,6 +728,26 @@ Status LgammaGrad(const Scope& scope, const Operation& op,
...
 
+Status SelectGrad(const Scope& scope, const Operation& op,
+                  const std::vector<Output>& grad_inputs,
+                  std::vector<Output>* grad_outputs) {
+  auto comparator = op.input(0);
+  auto x = op.input(1);
+  auto zeros = ZerosLike(scope, x);
+  auto grad = grad_inputs[0];
+
+  Scope grad_scope = scope.WithControlDependencies(grad);
```
Comment (suharshs) [inline tensorflow/cc/gradients/math_grad.cc]: Why do you need this control dependency? Isn't there a data dependency with grad already?

File: tensorflow/cc/gradients/math_grad_test.cc [hunk 2]
```diff
@@ -865,5 +865,14 @@ TEST_F(NaryGradTest, Minimum) {
   RunTest(x, x_init_value, y, shape);
 }
 
+TEST_F(NaryGradTest, Select) {
+  TensorShape x_shape({3, 4});
```
Comment (suharshs) [inline tensorflow/cc/gradients/math_grad_test.cc]: Nit: Since x_shape is the same as y_shape, maybe rename one to 'shape' and use that everywhere?

**Review by suharshs — CHANGES_REQUESTED** [review]
a couple minor comments. Thanks!

**Comment by facaiy** [comment]
Welcome back, @suharshs . I have removed grad_scope, x_shape, y_shape as requested.

**Review by suharshs — APPROVED** [review]

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by gunan** [comment]
Correction, this is the cause of the flakes. reverting this PR instead.

File: tensorflow/cc/gradients/math_grad_test.cc [hunk 3]
```diff
@@ -865,5 +865,13 @@ TEST_F(NaryGradTest, Minimum) {
   RunTest(x, x_init_value, y, shape);
 }
 
+TEST_F(NaryGradTest, Select) {
+  TensorShape shape({3, 4});
+  auto x1 = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
+  auto x2 = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
+  auto y = Where3(scope_, Greater(scope_, x1, x2), x1, x2);
```
Comment (facaiy) [inline tensorflow/cc/gradients/math_grad_test.cc]: @gunan @drpngx There seems something wrong with the condition `x1 > x2`, which might be unstable for gradient calculation.

Comment (facaiy) [reply tensorflow/cc/gradients/math_grad_test.cc]: @drpngx I want to confirm that:  Is `NaryGradTest.Select` the only one failed test?

Comment (drpngx) [reply tensorflow/cc/gradients/math_grad_test.cc]: I can't remember on the top of my head. I thought there was another one. You can try `bazel test --runs_per_test=100` and you should find it.

Comment (facaiy) [reply tensorflow/cc/gradients/math_grad_test.cc]: Fine, I'll look into it later. Thanks for your quick reply, @drpngx .

Comment (facaiy) [reply tensorflow/cc/gradients/math_grad_test.cc]: I have opened #15807 to fix the unstable test case. Might you take a look? Thanks.

**Pull request merged**